tls/s2n_crl.c

```diff
@@ -91,6 +91,47 @@ int s2n_crl_get_issuer_hash(struct s2n_crl *crl, uint64_t *hash)
     return S2N_SUCCESS;
 }
 
+int s2n_crl_validate_this_update(struct s2n_crl *crl)
+{
+    POSIX_ENSURE_REF(crl);
+    POSIX_ENSURE_REF(crl->crl);
+
+    ASN1_TIME *this_update = X509_CRL_get_lastUpdate(crl->crl);
+    POSIX_ENSURE_REF(this_update);
+
+    int ret = X509_cmp_time(this_update, NULL);
+    if (ret == 0) {
+        POSIX_BAIL(S2N_ERR_CRL_INVALID_THIS_UPDATE);
+    }
+    if (ret > 0) {
+        POSIX_BAIL(S2N_ERR_CRL_NOT_YET_VALID);
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_crl_validate_next_update(struct s2n_crl *crl)
```

Comment:
Should we provide a way for customers to distinguish between "expired" and other errors? Right now, I think they only have the error message, which you can't really act on. But maybe we can assume they'd always treat errors and "expired" the same?

Reply:
Yeah, that's a good point. Maybe we could return an S2N_RESULT and also set an `expired` bool that's passed in?

However, from the customer's perspective, I think it might make sense as is. If this function errors, the CRL is either expired, or there is no way to determine if it's expired or not. So, if the customer is calling validate to see if a new CRL should be fetched from the filesystem store, they would want to do that on any error. Additionally, if the customer is calling validate because they want to fail the handshake if their CRL has expired, it seems like they would also want the handshake to fail if they couldn't be sure whether it has expired or not.

Reply:
Related to this, the X509_verify_cert call should also ignore timestamp errors, since they're checked in these validate functions: 683da089a92f0ff93a4abd77492c6db6f0f65938

Reply:
I think you're right. And customers will at least have error messages for their logging.